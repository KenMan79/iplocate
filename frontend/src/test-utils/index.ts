export * from "./test-utils";
export * from "./ipdata";
