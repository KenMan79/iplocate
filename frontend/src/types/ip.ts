export interface IP {
  address: string;
  city: string;
  latitude: number;
  longitude: number;
}
