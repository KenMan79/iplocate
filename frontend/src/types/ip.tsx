export interface IP {
  address: string;
  latitude: string;
  longitude: string;
  city: string;
}
