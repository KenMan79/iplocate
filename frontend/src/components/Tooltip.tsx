import React, { useMemo } from "react";
import styled from "styled-components";
import { IP } from "types";
import { Card as C } from "./Card";

interface Props extends Partial<Omit<JSX.IntrinsicElements["div"], "ref">> {
  ip: IP;
}

export const Tooltip: React.FC<Props> = (props) => {
  const { ip, ...rest } = props;

  const data = useMemo(() => {
    const ret = [];

    const location = [
      ip.city.names?.en ?? "",
      ip.country.names?.en ?? "",
      ip.continent.names?.en ?? "",
    ]
      .filter(Boolean)
      .join(", ");

    const cidr = ip.traits.network ? ip.traits.network.split("/")[1] : "";
    const address = `${ip.traits.ipAddress}${cidr ? "/" : ""}${cidr}`;
    if (address) {
      ret.push({
        label: "IP",
        value: address,
      });
    }
    if (location) {
      ret.push({
        label: "Location",
        value: location,
      });
    }
    if (ip.location.accuracyRadius) {
      ret.push({
        label: "Accuracy",
        value: `${ip.location.accuracyRadius}km`,
      });
    }
    if (ip.location.timeZone) {
      ret.push({
        label: "Time Zone",
        value: ip.location.timeZone,
      });
    }
    return ret;
  }, [ip]);

  return (
    <Card {...rest}>
      <Details>
        {data.map(({ label, value }) => (
          <React.Fragment key={label}>
            <Label>{label}</Label>
            <Data>{value}</Data>
          </React.Fragment>
        ))}
      </Details>
    </Card>
  );
};

const Card = styled(C)`
  position: absolute;
  z-index: 50;
  min-width: 400px;
  box-shadow: 0px 3px 3px -2px rgba(0, 0, 0, 0.2),
    0px 3px 4px 0px rgba(0, 0, 0, 0.14), 0px 1px 8px 0px rgba(0, 0, 0, 0.12);
`;

const Label = styled.dt`
  display: block;
  margin: 0;
  padding-bottom: 0.5rem;
  width: 100%;
  font-weight: 600;
`;

const Data = styled.dd`
  display: block;
  margin: 0;
  padding-bottom: 1rem;
  width: 100%;
`;

const Details = styled.dl`
  margin: 0;
  padding: 0;
  color: var(--text-primary);

  ${Data}:last-of-type {
    padding-bottom: 0;
  }
`;
