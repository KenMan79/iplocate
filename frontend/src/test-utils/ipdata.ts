import { IP } from "types";

export const baseIP: IP = {
  continent: {
    code: "EU",
    geonameId: 6255148,
    names: {
      en: "Europe",
    },
  },
  country: {
    geonameId: 3017382,
    isInEuropeanUnion: true,
    isoCode: "FR",
    names: {
      en: "France",
    },
  },
  traits: {
    isAnonymous: false,
    isAnonymousProxy: false,
    isAnonymousVpn: false,
    isHostingProvider: false,
    isLegitimateProxy: false,
    isPublicProxy: false,
    isResidentialProxy: false,
    isSatelliteProvider: false,
    isTorExitNode: false,
    ipAddress: "2.2.2.2",
    network: "2.2.0.0/18",
  },
  city: {},
  location: {
    accuracyRadius: 500,
    latitude: 48.8582,
    longitude: 2.3387,
    timeZone: "Europe/Paris",
  },
};
