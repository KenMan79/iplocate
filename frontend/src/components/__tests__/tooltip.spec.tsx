import React from "react";
import { render, baseIP } from "test-utils";
import { IP } from "types";
import { Tooltip } from "../Tooltip";

it("matches snapshot", () => {
  const { asFragment } = render(<Tooltip ip={baseIP} />);
  expect(asFragment()).toMatchSnapshot();
});

it("builds location comma-separated", () => {
  const ip: IP = {
    ...baseIP,
    country: { names: { en: "Orangeland" } },
    city: { names: { en: "City" } },
    continent: { names: { en: "Antarctica" } },
  };
  const { getByText } = render(<Tooltip ip={ip} />);
  expect(getByText("City, Orangeland, Antarctica")).toBeInTheDocument();
});

it("ignores empty location values", () => {
  const ip: IP = {
    ...baseIP,
    country: { names: { en: "" } },
    city: { names: { en: "City" } },
    continent: { names: { en: "Antarctica" } },
  };
  const { getByText } = render(<Tooltip ip={ip} />);
  expect(getByText("City, Antarctica")).toBeInTheDocument();
});

it("ignores undefined location values", () => {
  const ip: IP = {
    ...baseIP,
    country: { names: { en: undefined! } },
    city: { names: { en: undefined! } },
    continent: { names: { en: undefined! } },
  };
  const { queryByText } = render(<Tooltip ip={ip} />);
  expect(queryByText("Location")).not.toBeInTheDocument();
});

it("builds ip/cidr from adderss and network", () => {
  const ip: IP = {
    ...baseIP,
    traits: {
      ...baseIP.traits,
      ipAddress: "1.2.3.4",
      network: "0.0.0.0/19",
    },
  };
  const { getByText } = render(<Tooltip ip={ip} />);
  expect(getByText("1.2.3.4/19")).toBeInTheDocument();
});

it("handles empty network cidr", () => {
  const ip: IP = {
    ...baseIP,
    traits: {
      ...baseIP.traits,
      ipAddress: "1.2.3.4",
      network: "",
    },
  };
  const { getByText } = render(<Tooltip ip={ip} />);
  expect(getByText("1.2.3.4")).toBeInTheDocument();
});

it("handles no network info", () => {
  const ip: IP = {
    ...baseIP,
    traits: {
      ...baseIP.traits,
      ipAddress: "",
      network: "",
    },
  };
  const { queryByText } = render(<Tooltip ip={ip} />);
  expect(queryByText("Address")).not.toBeInTheDocument();
});

it("adds accuracy radius in km if available", () => {
  const ip: IP = {
    ...baseIP,
    location: {
      ...baseIP.location,
      accuracyRadius: 42,
    },
  };
  const { getByText } = render(<Tooltip ip={ip} />);
  expect(getByText("42km")).toBeInTheDocument();
});

it("excludes accuracy radius if unavailable", () => {
  const { accuracyRadius, ...rest } = baseIP.location;
  const ip: IP = {
    ...baseIP,
    location: {
      ...baseIP.location,
      accuracyRadius: undefined!,
    },
  };
  const { queryByText } = render(<Tooltip ip={ip} />);
  expect(queryByText("Accuracy")).not.toBeInTheDocument();
});

it("adds time zone if available", () => {
  const ip: IP = {
    ...baseIP,
    location: {
      ...baseIP.location,
      timeZone: "sometimezonedata",
    },
  };
  const { getByText } = render(<Tooltip ip={ip} />);
  expect(getByText("Time Zone")).toBeInTheDocument();
  expect(getByText("sometimezonedata")).toBeInTheDocument();
});

it("excludes accuracy radius if unavailable", () => {
  const { accuracyRadius, ...rest } = baseIP.location;
  const ip: IP = {
    ...baseIP,
    location: {
      ...baseIP.location,
      timeZone: undefined!,
    },
  };
  const { queryByText } = render(<Tooltip ip={ip} />);
  expect(queryByText("Time Zone")).not.toBeInTheDocument();
});
