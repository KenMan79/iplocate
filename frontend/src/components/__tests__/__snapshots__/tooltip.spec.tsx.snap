// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  .c0 {
  padding: 2rem 1rem;
  margin: 0;
  border-radius: 5px;
  background-color: var(--background-card);
}

.c1 {
  position: absolute;
  z-index: 50;
  min-width: 400px;
  box-shadow: 0px 3px 3px -2px rgba(0,0,0,0.2), 0px 3px 4px 0px rgba(0,0,0,0.14),0px 1px 8px 0px rgba(0,0,0,0.12);
}

.c3 {
  display: block;
  margin: 0;
  padding-bottom: 0.5rem;
  width: 100%;
  font-weight: 600;
}

.c5 {
  display: block;
  margin: 0;
  padding-bottom: 1rem;
  width: 100%;
}

.c2 {
  margin: 0;
  padding: 0;
  color: var(--text-primary);
}

.c2 .c4:last-of-type {
  padding-bottom: 0;
}

<div
    class="c0 c1"
  >
    <dl
      class="c2"
    >
      <dt
        class="c3"
      >
        IP
      </dt>
      <dd
        class="c4 c5"
      >
        2.2.2.2/18
      </dd>
      <dt
        class="c3"
      >
        Location
      </dt>
      <dd
        class="c4 c5"
      >
        France, Europe
      </dd>
      <dt
        class="c3"
      >
        Accuracy
      </dt>
      <dd
        class="c4 c5"
      >
        500km
      </dd>
      <dt
        class="c3"
      >
        Time Zone
      </dt>
      <dd
        class="c4 c5"
      >
        Europe/Paris
      </dd>
    </dl>
  </div>
</DocumentFragment>
`;
